### PR TITLE
[Doris On ES] Add field context for string field keyword type

### PR DESCRIPTION
**Pull request by wuyunfeng**
https://github.com/apache/incubator-doris/issues/3304

This PR is just a transitional way，but it is better to move the predicates transformation from Doris BE to Doris BE, in this way, Doris BE is responsible for fetching data from ES.

 Add a  `enable_keyword_sniff ` configuration item in creating External Elasticsearch Table ，it default to true , would to sniff the `keyword` type on the `text analyzed` Field and return the `json_path` which substitute the origin col name.

```
CREATE EXTERNAL TABLE `test` (
  `k1` varchar(20) COMMENT "",
  `create_time` datetime COMMENT ""
) ENGINE=ELASTICSEARCH
PROPERTIES (
"hosts" = "http://10.74.167.16:8200",
"user" = "root",
"password" = "root",
"index" = "test",
"type" = "doc",
"enable_keyword_sniff" = "true"
);
```
note: `enable_keyword_sniff` default to  "true"

run this SQL：

```
select * from test where k1 = "wu yun feng"
```
 Output predicate DSL：

```
{"term":{"k1.keyword":"wu yun feng"}}
```
and in this PR, I remove the elasticsearch version detected logic for now this is useless, maybe future is needed.

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/planner/EsScanNode.java [hunk 1]
```diff
@@ -123,6 +123,9 @@ protected void toThrift(TPlanNode msg) {
         if (table.isDocValueScanEnable()) {
             esScanNode.setDocvalue_context(table.docValueContext());
         }
+        if (table.isKeywordSniffEnable() && table.fieldsContext().size() > 0) {
```
Comment (stalary) [inline fe/src/main/java/org/apache/doris/planner/EsScanNode.java:126]: !table.fieldsContext().isEmpty()是不是更好一点

Comment (wuyunfeng) [reply fe/src/main/java/org/apache/doris/planner/EsScanNode.java:126]: 这两个是一样的，第一个条件是用户主动配置的，第二个目前代码都是这么写的^o^

**Review by stalary — DISMISSED** [review]
LGTM

File: be/src/exec/es/es_predicate.h [hunk 2]
```diff
@@ -198,6 +198,10 @@ class EsPredicate {
         return _es_query_status;
     }
 
+    void field_context(const std::map<std::string, std::string>& field_context) {
```
Comment (morningman) [inline be/src/exec/es/es_predicate.h:201]: better to call `set_ field_context`

Comment (morningman) [reply be/src/exec/es/es_predicate.h:201]: And your UT failed.

**Comment by morningman** [comment]
> #3304
> 
> 这个方案其实是一个过度的解决方案，最佳的解决方案应该是将对EsScanNode的谓词转换在FE中完成，BE只负责扫描，这样将对查询的处理和执行完全分开。
> 
> 建表时增加一个 `enable_keyword_sniff ` 配置，默认是 "true", 会对分词字段进行探测获取`keyword`类型字段的`json path`.
> 
> ```
> 
> CREATE EXTERNAL TABLE `test` (
>   `k1` varchar(20) COMMENT "",
>   `create_time` datetime COMMENT ""
> ) ENGINE=ELASTICSEARCH
> PROPERTIES (
> "hosts" = "http://10.74.167.16:8200",
> "user" = "root",
> "password" = "root",
> "index" = "test",
> "type" = "doc",
> "enable_keyword_sniff" = "true"
> );
> ```
> 
> 其中 `enable_keyword_sniff` 默认就是true
> 
> 效果：
> 
> ```
> select * from test where k1 = "wu yun feng"
> ```
> 
> 输出：
> 
> ```
> {"term":{"k1.keyword":"wu yun feng"}}
> ```
> 
> 在这个PR中，我暂时把对ES的version检测移除了
> 
> 未来几个月将会将谓词的转换全部上移到FE中，处理这种就变得非常简单了。

Better in English

File: be/src/exec/es/es_predicate.cpp [hunk 3]
```diff
@@ -298,8 +302,12 @@ Status EsPredicate::build_disjuncts_list(const Expr* conjunct) {
         } else {
             is_not_null = true;
         }
+        std::string col = slot_desc->col_name();
```
Comment (kangkaisen) [inline be/src/exec/es/es_predicate.cpp:305]: Repeat three times.
```
        std::string col = slot_desc->col_name();
        if (_field_context.find(col) != _field_context.end()) {
            col = _field_context[col];
        }
```

Comment (wuyunfeng) [reply be/src/exec/es/es_predicate.cpp:305]: maybe instroduce a common function would increase the complexity

**Comment by stalary** [comment]
How is the review going? I need this feature. @morningman

File: gensrc/thrift/PlanNodes.thrift [hunk 4]
```diff
@@ -222,6 +222,18 @@ struct TEsScanNode {
...
+    //    "type": "text",
+    //    "fields": {
+    //        "keyword": {
+    //            "type": "keyword",
+    //            "ignore_above": 256
+    //           }
+    //    }
+    // }
+    // k1 > 'abc' -> k1.keyword > 'abc'
+    4: optional map<string, string> fields_context
```
Comment (imay) [inline gensrc/thrift/PlanNodes.thrift:236]: If this is only used for keyword path, `fields_context` seems no relation with keyword

File: fe/src/main/java/org/apache/doris/external/EsStateStore.java [hunk 5]
```diff
@@ -209,36 +206,54 @@ public EsTableState parseClusterState55(String responseStr, EsTable esTable)
...
-                                continue;
-                            }
-                            if (innerTypeObject.has("doc_values")) {
-                                boolean docValue = innerTypeObject.getBoolean("doc_values");
-                                if (docValue) {
-                                    esTable.addDocValueField(colName, colName);
+                // string-type field used keyword type to generate predicate
+                if (esTable.isKeywordSniffEnable()) {
+                    // if text field type seen, we should use the `field` keyword type?
+                    if ("text".equals(fieldType)) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/external/EsStateStore.java:212]: ignorecase compare?

Comment (wuyunfeng) [reply fe/src/main/java/org/apache/doris/external/EsStateStore.java:212]: Elasticsearch's type must be lowercase...

File: gensrc/thrift/PlanNodes.thrift (hunk 4 above)
Comment (wuyunfeng) [reply gensrc/thrift/PlanNodes.thrift:236]: https://www.elastic.co/guide/en/elasticsearch/reference/current/multi-fields.html

fields is what it means to be, maybe enable_keyword_sniff is no t suitable?

Comment (imay) [reply gensrc/thrift/PlanNodes.thrift:236]: You're right

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**